packages/@glimmer/application/src/components/component-managers/custom.ts

```diff
@@ -0,0 +1,375 @@
+    if (hasDestructors(state.delegate)) {
+      return state;
+    } else {
+      return null;
+    }
+  }
+
+  getCapabilities({
+    delegate,
+  }: CustomComponentDefinitionState<ComponentInstance>): ComponentCapabilities {
+    return Object.assign({}, CAPABILITIES, {
+      updateHook: delegate.capabilities.updateHook,
+    });
+  }
+
+  getTag({ args }: CustomComponentState<ComponentInstance>): Tag {
+    if (isConst(args)) {
+      // returning a const tag skips the update hook (VM BUG?)
+      return createTag();
+    } else {
+      return args.tag;
+    }
+  }
+
+  didRenderLayout({ delegate, component }: CustomComponentState<ComponentInstance>, bounds: VMBounds) {
+    if (hasDidRenderLayout(delegate)) {
+      delegate.__glimmer__didRenderLayout(component, new Bounds(bounds));
+    }
+  }
+
+  didUpdateLayout() {}
+
+  getJitStaticLayout({ definition }: CustomComponentDefinitionState<ComponentInstance>): CompilableProgram {
+    return definition.template.asLayout();
+  }
+
+  getAotStaticLayout({ definition }: CustomComponentDefinitionState<ComponentInstance>): Invocation {
+    const { handle, symbolTable } = definition;
+    return {
+      handle,
+      symbolTable
+    };
+  }
+}
+
+/**
+ * Stores internal state about a component instance after it's been created.
+ */
+export class CustomComponentState<ComponentInstance> {
+  constructor(
+    public delegate: ManagerDelegate<ComponentInstance>,
+    public component: ComponentInstance,
+    public args: CapturedArguments,
+    public namedArgsProxy?: {}
+  ) {}
+
+  destroy() {
+    const { delegate, component } = this;
+
+    if (hasDestructors(delegate)) {
+      delegate.destroyComponent(component);
+    }
+  }
+}
+
+export interface CustomComponentDefinitionState<ComponentInstance> {
+  name: string;
+  delegate: ManagerDelegate<ComponentInstance>;
+  ComponentClass: ComponentFactory;
+  definition: CustomComponentDefinition<ComponentInstance>;
+}
+
+const CUSTOM_COMPONENT_MANAGER = new CustomComponentManager();
+
+export class CustomComponentDefinition<ComponentInstance> implements AotComponentDefinition, JitComponentDefinition {
+  public state: CustomComponentDefinitionState<ComponentInstance>;
+  public manager = CUSTOM_COMPONENT_MANAGER as CustomComponentManager<ComponentInstance>;
+  public handle: number;
+  public symbolTable: ProgramSymbolTable;
+  public template: Template;
+
+  constructor(
+    name: string,
+    ComponentClass: ComponentFactory,
+    delegate: ManagerDelegate<ComponentInstance>,
+    templateOrHandle: Template | number,
+    symbolTable?: ProgramSymbolTable
+  ) {
+    if (typeof templateOrHandle === 'number') {
+      this.handle = templateOrHandle;
+      this.symbolTable = symbolTable;
+    } else {
+      this.template = templateOrHandle;
+    }
+
+    this.state = {
+      name,
+      ComponentClass,
+      delegate,
+      definition: this
+    };
+  }
+}
+
+export interface ComponentFactory<C = {}> {
+  class: C;
+}
```

Comment:
Newline: you really should turn on the vscode setting 😄

Reply:
Really it should be a lint. 😁 My hard drive crashed and I had not re-enabled it since reinstalling VS Code.